google/cloud/forseti/common/util/rule_validator.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2017 The Forseti Security Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Validation routines for GCE firewall rules."""
+from __future__ import unicode_literals
+
+import re
+
+# Allowed items in a firewall rule.
+ALLOWED_RULE_ITEMS = frozenset(('allowed', 'denied', 'description', 'direction',
+                                'disabled', 'name', 'network', 'priority',
+                                'sourceRanges', 'destinationRanges',
+                                'sourceTags', 'targetTags', 'logConfig'))
+
+# Top level required items for a valid rule.
+REQUIRED_RULE_ITEMS = frozenset(('name', 'network'))
+
+# Keys that allow a list but only up to 256 entries.
+MAX_256_VALUE_KEYS = frozenset(
+    ['sourceRanges', 'sourceTags', 'targetTags', 'destinationRanges'])
+
+# Name restrictions described at
+# https://cloud.google.com/compute/docs/reference/rest/v1/firewalls
+VALID_RESOURCE_NAME_RE = re.compile('(?:^[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?$)')
+
+
+def validate_gcp_rule(rule):
```

Comment:
Consider having this method raise the InvalidFirewallRuleError exception instead of passing the exception message back to the add_rules method.

Reply:
returning a string is friendlier to outside users of the library as opposed to building try/catch wrappers when validating a large number of rules.

Reply:
Understood, thanks.